SSA/Projects/CIRCT/DCxComb/DCxComb.lean

```diff
@@ -0,0 +1,588 @@
+| Ty.valuetokenstream w => CIRCTStream.DCxCombOp.ValueStream (BitVec w) × CIRCTStream.DCxCombOp.TokenStream
+| .hList l => HVector BitVec l -- het list of bitvec whose lengths are contained in l
+| .icmpPred => CIRCTStream.DCxCombOp.IcmpPredicate
+
+def HVector.replicateToList {α : Type} {f : α → Type} {a : α} :
+    {n : Nat} → HVector f (List.replicate n a) → List (f a)
+  | 0, _ => []
+  | n+1, HVector.cons x xs => x :: replicateToList xs
+
+def ofString? (s : String) : Option CIRCTStream.DCxCombOp.IcmpPredicate :=
+  match s with
+  | "eq" => some .eq
+  | "ne" => some .ne
+  | "slt" => some .slt
+  | "sle" => some .sle
+  | "sgt" => some .sgt
+  | "sge" => some .sge
+  | "ult" => some .ult
+  | "ule" => some .ule
+  | "ugt" => some .ugt
+  | "uge" => some .uge
+  | _     => none
+
+
+def_denote for DCxComb where
+  | .fst => fun s => s.fst
+  | .fstVal _ => fun s => s.fst
+  | .snd => fun s => s.snd
+  | .sndVal _ => fun s => s.snd
+  | .merge => fun s₁ s₂ => CIRCTStream.DCxCombOp.merge s₁ s₂
+  | .branch => fun s => CIRCTStream.DCxCombOp.branch s
+  | .fork => fun s => CIRCTStream.DCxCombOp.fork s
+  | .join _ => fun xs => CIRCTStream.DCxCombOp.join (HVector.replicateToList (f := TyDenote.toType) xs)
+  | .select => fun s₁ s₂ c => CIRCTStream.DCxCombOp.select s₁ s₂ c
+  | .sink => fun s => CIRCTStream.DCxCombOp.sink s
+  | .source => fun s => CIRCTStream.DCxCombOp.source s
+  | .pack _ => fun s₁ s₂ => CIRCTStream.DCxCombOp.pack s₁ s₂
+  | .pair _ => fun s₁ s₂ => (s₁, s₂)
+  | .unpack _ => fun s => CIRCTStream.DCxCombOp.unpack s
+  | .popReady _ n => fun s => CIRCTStream.DCxCombOp.popReady s n
+  | .pushReady _ n => fun s => CIRCTStream.DCxCombOp.pushReady s n
+  | .add _ _ => fun xs => CIRCTStream.DCxCombOp.add (HVector.replicateToList (f := TyDenote.toType) xs)
+  | .and _ _ => fun xs => CIRCTStream.DCxCombOp.and (HVector.replicateToList (f := TyDenote.toType) xs)
+  | .concat _ => fun xs => CIRCTStream.DCxCombOp.concat xs
+  | .divs _ => BitVec.sdiv
+  | .divu _ => BitVec.udiv
+  | .extract _ _ => fun x => CIRCTStream.DCxCombOp.extract x _
+  | .icmp p _ => fun x y => CIRCTStream.DCxCombOp.icmp (Option.get! (ofString? p)) x y
+  | .mods _ => BitVec.smod
+  | .modu _ => BitVec.umod
+  | .mul _ _ => fun xs => CIRCTStream.DCxCombOp.mul (HVector.replicateToList (f := TyDenote.toType) xs)
+  | .mux _ => fun x y => CIRCTStream.DCxCombOp.mux x y
+  | .or _ _ => fun xs => CIRCTStream.DCxCombOp.or (HVector.replicateToList (f := TyDenote.toType) xs)
+  | .parity _ => fun x => CIRCTStream.DCxCombOp.parity x
+  | .replicate _ n => fun xs => CIRCTStream.DCxCombOp.replicate xs n
+  | .shl _ => fun x y => CIRCTStream.DCxCombOp.shl x y
+  | .shrs _ => BitVec.sshiftRight'
+  | .shru _ => fun x y => CIRCTStream.DCxCombOp.shru x y
+  | .sub _ => BitVec.sub
+  | .xor _ _ => fun xs => CIRCTStream.DCxCombOp.xor (HVector.replicateToList (f := TyDenote.toType) xs)
+
+
+end Dialect
+
+
+def mkTy : MLIR.AST.MLIRType φ → MLIR.AST.ExceptM DCxComb DCxComb.Ty
```

Comment:
```suggestion
def mkTy : MLIR.AST.MLIRType 0 → MLIR.AST.ExceptM DCxComb DCxComb.Ty
```

The `φ` gives a static bound on the number of mvars. By setting it zero you don't have to check the mvar branch below!